### PR TITLE
Prestwich/super pending

### PR DESCRIPTION
**Pull request by prestwich**
## Motivation

Provide a reasonable and working alternative to gas escalator middleware. 

## Solution

- Introduce a new Future `EscalatingPending` that broadcasts a series of transactions
    - functions similarly to `PendingTransaction` but with a simpler (but less informative) state machine
    - it returns the _first_ receipt to confirm.
    - it also returns the _first_ RPC error to occur
    - it allows users to configure polling and broadcast intervals
- Add a middleware method `send_escalating` 
    - allows users to instantiate with a TypedTransaction and an escalation policy
    - may be better as a middleware extension trait?

## things I don't like
- Invariant enforcement is done in the Middleware method rather than the constructor. 
    - This is brittle
    - it's hard to see how to sign and nonce correctly otherwise
- Naming

## PR Checklist

- [ ] Added Tests
- [X] Added Documentation
- [x] Updated the changelog

### CODE REVIEW

File: ethers-providers/src/pending_escalator.rs
Comment (ltchang2019) [inline ethers-providers/src/pending_escalator.rs:151]: So if we have already broadcast the most escalated tx, we will just wait on that one indefinitely?

File: ethers-providers/src/pending_escalator.rs
Comment (ltchang2019) [inline ethers-providers/src/pending_escalator.rs:181]: Is this `cx.waker().wake_by_ref()` necessary? Won't executor automatically put the EscalatingPending future back in queue? Other paths here that return Poll::Pending don't call wake

Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:181]: the goal is to explicitly notify the executor that this is ready to be polled again _immediately_. i.e. override the executor's scheduling and jump to the front of the queue as much as we have the ability to, as we know that there's no async operation happening to wait on

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:151]: yes

File: ethers-providers/src/lib.rs
Comment (gakonst) [inline ethers-providers/src/lib.rs:302]: You can use `self.default_sender` instead of `from`

File: ethers-providers/src/lib.rs
Comment (gakonst) [inline ethers-providers/src/lib.rs:303]: Maybe this should be a reference?

File: ethers-providers/src/lib.rs
Comment (gakonst) [inline ethers-providers/src/lib.rs:309]: Should the range be part of `EscalationPolicy`?

File: ethers-providers/src/lib.rs
Comment (gakonst) [inline ethers-providers/src/lib.rs:317]: Is this really bad UX? Should we change it to something better?

**Review by gakonst — COMMENTED** [review]
some minor qs, love the direction

File: ethers-providers/src/lib.rs
Comment (prestwich) [reply ethers-providers/src/lib.rs:303]: I'm considering making this a trait, as we really want something that generates an arbitrary number txns instead of always 5. we can't do an unbounded number due to the constraints on pre-signing before instantiating the escalating tho

File: ethers-providers/src/lib.rs
Comment (prestwich) [reply ethers-providers/src/lib.rs:302]: `default_sender()` returns `Option<H160>` what should be the behavior in the `None` case?

File: ethers-providers/src/lib.rs
Comment (prestwich) [reply ethers-providers/src/lib.rs:317]: a shortcut function for signing transactions might be good?

File: ethers-providers/src/lib.rs
Comment (prestwich) [reply ethers-providers/src/lib.rs:309]: yes, but haven't gotten there yet

File: ethers-providers/src/lib.rs
Comment (prestwich) [reply ethers-providers/src/lib.rs:302]: I guess `unwrap_or_default`. that'll cause clear errors when we join_all later

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [inline ethers-providers/src/pending_escalator.rs:139]: @gakonst please check this pattern out, as I am unclear why we use `Delay` futures elsewhere. is there some significant behavior difference I'm missing?

Comment (mattsse) [reply ethers-providers/src/pending_escalator.rs:139]: the futures_timer Delay type uses an additional global timerhandle thread that makes sure the delayed task gets woken up at the right time, however, futures_timer hasn't been updated in a while there's also tokio::time::Sleep now, which I guess works similarly but probably uses some Runtime Handle instead.

looking at the Sleeping state, there might be a case were the Future returns Pending and gets not woken up again (elpased < polling_interval)

File: ethers-providers/src/pending_escalator.rs
Comment (mattsse) [reply ethers-providers/src/pending_escalator.rs:176]: this section looks like it could be wrapped in a loop over all futures? So that `Poll::Ready(Ok(None)) =>` merely results in a continue?
Is the order in which they should be polled here important, if not then `Vec::swap_remove` could be useful.

**Review by mattsse — COMMENTED** [review]
I feel like this could even be two different types.
An additional `Stream` type that processes all transactions.
The `EscalatingPending` Future that polls the underlying stream and returns the first value?

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:139]: `tokio::time::Sleep` won't work in WASM iirc

I believe all common executors schedule futures for polling in frames regardless of whether the waker has explicitly requested waking

Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:139]: didn't finish the previous thought before submitting the tx

given that this could cause issues with uncommon executors, should we audit all our futures to ensure that all polls either request rewaking _or_ poll another future?

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:176]: I mean it kinda is right now, it's just ensuring that it allows the executor to schedule each step of the loop and interleave other work

**Comment by prestwich** [comment]
test failure is unrelated to these changes

@gakonst @mattsse do you think that this belongs on the main middleware or on some sort of extension trait?

**Comment by mattsse** [comment]
> @gakonst @mattsse do you think that this belongs on the main middleware or on some sort of extension trait?

What, could be nice if this could be somehow integrated with the current `send_transaction -> PendingTx`, so that we can do something like `p.send_transaction(tx).escalate(||...).await` and `p.send_transaction(tx).await.await` still works.

I believe this would be possible with an additional future type (`SendTx` or something?) that would have `Output = PendingTransaction` and a `fn escalate`. But this would mean, `async fn send_transaction` would become `fn send_transaction() -> SendTx`.

this would be a breaking change however.

The easiest way would probably be to add another provider function, as is the case now.

**Comment by prestwich** [comment]
the problem with that is that the signing _must_ be done in the top-level middleware. It _can't_ be done by the provider that the pending transaction has access to. So that feature is blocked by an object-safe `Middleware` trait

**Comment by mattsse** [comment]
ah I see, makes sense, this would then be limited to only the SignerMiddleware...

**Comment by prestwich** [comment]
Not exactly, as the raw provider may sign in many cases (using eth_sign or similar). However, relying on the provider would result in a conflict _if_ any `SignerMiddleware` were used. 

E.g `SignerMiddleware( Provider )`.  The `SignerMiddleware` may be signing with address `0x1234...` where the node controls address `0xabcd...`. This would cause errors and other unexpected behavior if we choose to have signing in the `Provider` held by a `PendingTransaction`

File: ethers-providers/src/pending_escalator.rs
Comment (gakonst) [inline ethers-providers/src/pending_escalator.rs:81]:
```suggestion
        self.polling_interval = Duration::from_millis(duration);
```

File: ethers-providers/src/pending_escalator.rs
Comment (gakonst) [reply ethers-providers/src/pending_escalator.rs:190]: Looping over all futures is an anti-pattern I think, the state machine _may_ be slightly easier to write but as @prestwich says it's better to do it on a per-fut basis

**Review by gakonst — COMMENTED** [review]
looking good, I am OK with placing it as an extra function as-is. I also liked @mattsse's idea of adding an extra type on top of `PendingTransaction` so that we can do `send_transaction(tx).await?.escalator(...).await?` to escalate and `send_transaction(tx).await?.await?` for simple usage

File: ethers-providers/src/pending_escalator.rs
Comment (mattsse) [reply ethers-providers/src/pending_escalator.rs:190]: Isn't the `CheckingReceipts` state essentially a `FuturesOrdered` but we only poll one future at a time?

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:76]: going to change it to use `Into<Duration>` to match the behavior of `PendingTransaction`

**Comment by prestwich** [comment]
latest commit uses `FuturesUnordered`. I think this is better as @mattsse says. We _don't_ require ordering here, so this functions similarly to select_all, but is easier to manage

**Comment by prestwich** [comment]
I have no other outstanding changes so I'll update the changelog and this is ready for consideration

**Review by gakonst — APPROVED** [review]
I like it. Thank you for the hard work, we can iterate on it post-merge. @mattsse any further comments?

**Comment by mattsse** [comment]
> I like it. Thank you for the hard work, we can iterate on it post-merge. @mattsse any further comments?

lgtm

File: ethers-providers/src/pending_escalator.rs
Comment (prestwich) [reply ethers-providers/src/pending_escalator.rs:203]: ah outdated comment. I'll review a those real quick before merge

**Comment by gakonst** [comment]
gg. Should we receive the escalator middleware?

**Pull request merged**
